File: frontend/lib/services/personal_loan_service.dart
import 'dart:async';

import 'package:flutter/foundation.dart';

import '../models/personal_loan.dart';
import '../models/personal_loan_emi.dart';
import 'api_client.dart';

/// Personal loans, backed by the FastAPI `/personal-loans` endpoints. Persists
/// loans + monthly EMIs; each EMI is marked paid one at a time.
class PersonalLoanService extends ChangeNotifier {
  PersonalLoanService({ApiClient? client}) : _api = client ?? ApiClient();

  final ApiClient _api;
  final List<PersonalLoan> _loans = [];
  bool _loading = false;

  bool get loading => _loading;

  List<PersonalLoan> all() => List.unmodifiable(_loans);

  PersonalLoan? byId(String id) =>
      _loans.where((l) => l.id == id).cast<PersonalLoan?>().firstOrNull;

  Future<void> refresh() async {
    _loading = true;
    notifyListeners();
    try {
      final data = await _api.get('/personal-loans');
      final fresh = (data as List)
          .map((j) => _fromJson(j as Map<String, dynamic>))
          .toList();
      _loans
        ..clear()
        ..addAll(fresh);
    } catch (_) {
      // keep cache
    } finally {
      _loading = false;
      notifyListeners();
    }
  }

  /// Creates a personal loan (persists + builds the EMI schedule server-side).
  Future<PersonalLoan> create({
    required String vehicleNumber,
    int? financerId,
    required int loanAmount,
    required int emiAmount,
    required int tenureMonths,
    required DateTime loanDate,
    String? phone,
  }) async {
    final body = <String, dynamic>{
      'vehicle_number': vehicleNumber,
      'financer_id': financerId,
      'loan_amount': loanAmount,
      'emi_amount': emiAmount,
      'tenure_months': tenureMonths,
      'loan_date': _dateStr(loanDate),
      'phone': (phone ?? '').trim().isEmpty ? null : phone!.trim(),
    }..removeWhere((_, v) => v == null);
    final json = await _api.post('/personal-loans', body: body);
    final loan = _fromJson(json as Map<String, dynamic>);
    _loans.insert(0, loan);
    notifyListeners();
    return loan;
  }

  /// Marks one EMI paid. Optimistic, then persisted + reconciled by refresh.
  void markEmiPaid(String loanId, String emiId) {
    final loan = byId(loanId);
    final emi =
        loan?.emis.where((e) => e.id == emiId).cast<PersonalLoanEmi?>().firstOrNull;
    if (loan == null || emi == null) return;
    emi.status = 'paid';
    emi.paidDate = DateTime.now();
    if (loan.emis.every((e) => e.isPaid)) loan.loanStatus = 'closed';
    notifyListeners();
    unawaited(_api
        .post('/personal-loans/$loanId/emis/$emiId/pay')
        .then((_) => refresh())
        .catchError((_) => null));
  }

  void delete(String id) {
    _loans.removeWhere((l) => l.id == id);
    notifyListeners();
    unawaited(_api.delete('/personal-loans/$id').catchError((_) => null));
  }

  // ── JSON mapping ────────────────────────────────────────────────────────────
  static String _dateStr(DateTime d) => d.toIso8601String().split('T').first;

  static DateTime? _date(dynamic s) =>
      (s is String && s.isNotEmpty) ? DateTime.tryParse(s) : null;

  static int _int(dynamic v) =>
      v == null ? 0 : (num.tryParse(v.toString())?.round() ?? 0);

  PersonalLoan _fromJson(Map<String, dynamic> j) {
    return PersonalLoan(
      id: j['id'].toString(),
      vehicleNumber: (j['vehicle_number'] as String?) ?? '',
      financerId: (j['financer_id'] as num?)?.toInt(),
      financerName: j['financer_name'] as String?,
      loanAmount: _int(j['loan_amount']),
      emiAmount: _int(j['emi_amount']),
      tenureMonths: (j['tenure_months'] as num?)?.toInt() ?? 0,
      loanDate: _date(j['loan_date']) ?? DateTime.now(),
      firstDueDate: _date(j['first_due_date']) ?? DateTime.now(),
      phone: j['phone'] as String?,
      loanStatus: (j['loan_status'] as String?) ?? 'active',
      emis: (j['emis'] as List?)
              ?.map((e) => _emiFromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  PersonalLoanEmi _emiFromJson(Map<String, dynamic> j) {
    return PersonalLoanEmi(
      id: j['id'].toString(),
      sequenceNumber: (j['sequence_number'] as num).toInt(),
      dueDate: _date(j['due_date']) ?? DateTime.now(),
      amount: _int(j['amount']),
      status: (j['status'] as String?) == 'paid' ? 'paid' : 'pending',
      paidDate: _date(j['paid_date']),
    );
  }
}

File: backend/app/services/personal_loan_service.py
"""Business logic for personal loans (simple monthly EMI, mark-paid)."""
import calendar
from datetime import date, datetime, timezone

from fastapi import HTTPException
from sqlalchemy.orm import Session

from app.dao.personal_loan_dao import PersonalLoanDAO
from app.models.enums import InstallmentStatus
from app.models.personal_loan import PersonalLoan
from app.models.personal_loan_emi import PersonalLoanEmi
from app.models.personal_loan_financer import PersonalLoanFinancer
from app.schemas.personal_loan import PersonalLoanCreate, PersonalLoanFinancerCreate


def _add_months(d: date, n: int) -> date:
    m = d.month - 1 + n
    y = d.year + m // 12
    m = m % 12 + 1
    last = calendar.monthrange(y, m)[1]
    return date(y, m, min(d.day, last))


class PersonalLoanService:
    # ── financers (own master for this module) ───────────────────────────────
    @staticmethod
    def financers(db: Session) -> list[PersonalLoanFinancer]:
        return PersonalLoanDAO.financers(db)

    @staticmethod
    def create_financer(
        db: Session, data: PersonalLoanFinancerCreate
    ) -> PersonalLoanFinancer:
        financer = PersonalLoanFinancer(name=data.name.strip())
        PersonalLoanDAO.add_financer(db, financer)
        db.commit()
        return PersonalLoanDAO.financer(db, financer.id)

    @staticmethod
    def delete_financer(db: Session, financer_id: int) -> None:
        financer = PersonalLoanDAO.financer(db, financer_id)
        if financer is None:
            raise HTTPException(status_code=404, detail="Financer not found")
        PersonalLoanDAO.delete_financer(db, financer)
        db.commit()

    # ── loans ────────────────────────────────────────────────────────────────
    @staticmethod
    def list(db: Session) -> list[PersonalLoan]:
        return PersonalLoanDAO.loans(db)

    @staticmethod
    def get(db: Session, loan_id: int) -> PersonalLoan:
        loan = PersonalLoanDAO.get(db, loan_id)
        if loan is None:
            raise HTTPException(status_code=404, detail="Personal loan not found")
        return loan

    @staticmethod
    def create(db: Session, data: PersonalLoanCreate, *, created_by: int) -> PersonalLoan:
        if data.tenure_months <= 0:
            raise HTTPException(status_code=400, detail="Tenure must be at least 1 month")
        first_due = _add_months(data.loan_date, 1)
        loan = PersonalLoan(
            vehicle_number=data.vehicle_number.strip(),
            financer_id=data.financer_id,
            loan_amount=data.loan_amount,
            emi_amount=data.emi_amount,
            tenure_months=data.tenure_months,
            loan_date=data.loan_date,
            first_due_date=first_due,
            phone=(data.phone or "").strip() or None,
            loan_status="active",
            remarks=data.remarks,
            created_by=created_by,
        )
        for i in range(data.tenure_months):
            loan.emis.append(
                PersonalLoanEmi(
                    sequence_number=i + 1,
                    due_date=_add_months(data.loan_date, i + 1),
                    amount=data.emi_amount,
                    status=InstallmentStatus.pending,
                )
            )
        PersonalLoanDAO.add(db, loan)
        db.commit()
        return PersonalLoanDAO.get(db, loan.id)

    @staticmethod
    def mark_emi_paid(db: Session, loan_id: int, emi_id: int) -> PersonalLoan:
        loan = PersonalLoanService.get(db, loan_id)
        emi = PersonalLoanDAO.get_emi(db, emi_id)
        if emi is None or emi.personal_loan_id != loan.id:
            raise HTTPException(status_code=404, detail="EMI not found")
        emi.status = InstallmentStatus.paid
        emi.paid_date = date.today()
        if all(e.status == InstallmentStatus.paid for e in loan.emis):
            loan.loan_status = "closed"
            loan.closed_at = datetime.now(timezone.utc)
        db.commit()
        return PersonalLoanDAO.get(db, loan_id)

    @staticmethod
    def delete(db: Session, loan_id: int) -> None:
        loan = PersonalLoanService.get(db, loan_id)
        db.delete(loan)
        db.commit()
